open_spiel/games/othello.cc

```diff
@@ -0,0 +1,344 @@
+    /*provides_observation_tensor=*/true,
+    /*parameter_specification=*/{}  // no parameters
+};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params) {
+  return std::shared_ptr<const Game>(new OthelloGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+}  // namespace
+
+std::tuple<int, int> GetNext(int row, int col, Direction dir) {
+  switch (dir) {
+    case Direction::kUp:
+      return std::make_tuple(row - 1, col);
+    case Direction::kDown:
+      return std::make_tuple(row + 1, col);
+    case Direction::kLeft:
+      return std::make_tuple(row, col - 1);
+    case Direction::kRight:
+      return std::make_tuple(row, col + 1);
+    case Direction::kUpRight:
+      return std::make_tuple(row - 1, col + 1);
+    case Direction::kUpLeft:
+      return std::make_tuple(row - 1, col - 1);
+    case Direction::kDownRight:
+      return std::make_tuple(row + 1, col + 1);
+    case Direction::kDownLeft:
+      return std::make_tuple(row + 1, col - 1);
+    default:
+      SpielFatalError(absl::StrCat("Found unmatched case in GetNext."));
+  }
+}
+
+inline bool OthelloState::OnBoard(int row, int col) const {
+  return (0 <= row) && (row < kNumRows) && (0 <= col) && (col < kNumCols);
+}
+
+int OthelloState::CountSteps(Player player, int row, int col, Direction dir) const {
+  std::tuple<int, int> rowcol = GetNext(row, col, dir);
+  row = std::get<0>(rowcol);
+  col = std::get<1>(rowcol);
+
+  int count = 0;
+  CellState cell = PlayerToState(player);
+  while (OnBoard(row, col)) {
+    if (BoardAt(row, col) == cell) {
+      return count;
+    } else if (BoardAt(row, col) == CellState::kEmpty) {
+      return 0;
+    }
+      
+    count++;
+    rowcol = GetNext(row, col, dir);
+    row = std::get<0>(rowcol);
+    col = std::get<1>(rowcol);
+  }
+
+  return 0;
+}
+
+bool OthelloState::CanCapture(Player player, int move) const {
+  if (board_[move] != CellState::kEmpty) return false;
+
+  std::tuple<int, int> row_col = XYFromCode(move);
+  int row = std::get<0>(row_col);
+  int col = std::get<1>(row_col);
+
+  bool valid = false;
+  for (int direction = Direction::kUp; direction < Direction::kLast; direction++) {
+    if (CountSteps(player, row, col, static_cast<Direction>(direction)) != 0) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void OthelloState::Capture(Player player, int row, int col, Direction dir, int steps) {
+  std::tuple<int, int> rowcol = GetNext(row, col, dir);
+  row = std::get<0>(rowcol);
+  col = std::get<1>(rowcol);
+  CellState cell = PlayerToState(player);
+
+  for (int step = 0; step < steps; step++) {
+    if (BoardAt(row, col) == CellState::kEmpty || BoardAt(row, col) == cell) {
+      SpielFatalError(absl::StrCat("Cannot capture cell (", row, ", ", col, ")"));
+    }
+
+    board_[row * kNumCols + col] = cell;
+    
+    rowcol = GetNext(row, col, dir);
+    row = std::get<0>(rowcol);
+    col = std::get<1>(rowcol);
+  }
+}
+
+int OthelloState::DiskCount(Player player) const {
```

Comment:
Is this number worth caching?

Reply:
Good question. This can be done for performance reasons, and it is probably not too hard to do. You just need to update it when disks are captured. I can add this if we think it's worthwhile.